// vue-tsc/out/proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadTsLib = exports.createProgramProxy = void 0;
const ts = require("typescript/lib/tsserverlibrary");
const apis = require("./apis");
const vueTs = require("@volar/vue-typescript");
let projectVersion = 0;
function createProgramProxy(options, // rootNamesOrOptions: readonly string[] | CreateProgramOptions,
_options, _host, _oldProgram, _configFileParsingDiagnostics) {
    var _a, _b;
    if (!options.options.noEmit && !options.options.emitDeclarationOnly)
        return doThrow('js emit is not support');
    if (!options.host)
        return doThrow('!options.host');
    projectVersion++;
    const host = options.host;
    const vueCompilerOptions = getVueCompilerOptions();
    const scripts = new Map();
    const vueLsHost = Object.assign(Object.assign({}, host), { resolveModuleNames: undefined, writeFile: (fileName, content) => {
            if (fileName.indexOf('__VLS_') === -1) {
                host.writeFile(fileName, content, false);
            }
        }, getCompilationSettings: () => options.options, getVueCompilationSettings: () => vueCompilerOptions, getScriptFileNames: () => {
            return options.rootNames;
        }, getScriptVersion,
        getScriptSnapshot, getProjectVersion: () => {
            return projectVersion.toString();
        }, getProjectReferences: () => options.projectReferences, isTsc: true });
    const vueLsCtx = (_b = (_a = options.oldProgram) === null || _a === void 0 ? void 0 : _a.__VLS_vueCtx) !== null && _b !== void 0 ? _b : vueTs.createLanguageServiceContext(ts, vueLsHost);
    const proxyApis = apis.register(ts, vueLsCtx);
    const program = new Proxy({}, {
        get: (_, property) => {
            if (property in proxyApis) {
                return proxyApis[property];
            }
            return vueLsCtx.typescriptLanguageService.getProgram()[property];
        },
    });
    program.__VLS_vueCtx = vueLsCtx;
    for (const rootName of options.rootNames) {
        // register file watchers
        host.getSourceFile(rootName, ts.ScriptTarget.ESNext);
    }
    return program;
    function getVueCompilerOptions() {
        const tsConfig = options.options.configFilePath;
        if (typeof tsConfig === 'string') {
            return vueTs.tsShared.createParsedCommandLine(ts, ts.sys, tsConfig).vueOptions;
        }
        return {};
    }
    function getScriptVersion(fileName) {
        var _a, _b;
        return (_b = (_a = getScript(fileName)) === null || _a === void 0 ? void 0 : _a.version) !== null && _b !== void 0 ? _b : '';
    }
    function getScriptSnapshot(fileName) {
        var _a;
        return (_a = getScript(fileName)) === null || _a === void 0 ? void 0 : _a.scriptSnapshot;
    }
    function getScript(fileName) {
        var _a, _b, _c, _d, _e, _f;
        const script = scripts.get(fileName);
        if ((script === null || script === void 0 ? void 0 : script.projectVersion) === projectVersion) {
            return script;
        }
        const modifiedTime = (_d = (_c = (_b = (_a = ts.sys).getModifiedTime) === null || _b === void 0 ? void 0 : _b.call(_a, fileName)) === null || _c === void 0 ? void 0 : _c.valueOf()) !== null && _d !== void 0 ? _d : 0;
        if ((script === null || script === void 0 ? void 0 : script.modifiedTime) === modifiedTime) {
            return script;
        }
        if (host.fileExists(fileName)) {
            const fileContent = host.readFile(fileName);
            if (fileContent !== undefined) {
                const script = {
                    projectVersion,
                    modifiedTime,
                    scriptSnapshot: ts.ScriptSnapshot.fromString(fileContent),
                    version: (_f = (_e = host.createHash) === null || _e === void 0 ? void 0 : _e.call(host, fileContent)) !== null && _f !== void 0 ? _f : fileContent,
                };
                scripts.set(fileName, script);
                return script;
            }
        }
    }
}
exports.createProgramProxy = createProgramProxy;
function loadTsLib() {
    return ts;
}
exports.loadTsLib = loadTsLib;
function doThrow(msg) {
    console.error(msg);
    throw msg;
}
